proc/threads.c

```diff
@@ -1508,6 +1508,59 @@ void threads_setupUserReturn(void *retval, cpu_context_t *ctx)
 }
 
 
+int threads_sigsuspend(unsigned int mask)
+{
+	int ret;
+	thread_t *thread;
+	spinlock_ctx_t sc;
+	cpu_context_t *ctx, *signalCtx;
+	unsigned int oldmask;
+	void *f;
+	thread = proc_current();
+
+
+	void *const kstackTop = thread->kstack + thread->kstacksz;
+	ctx = kstackTop - sizeof(*ctx);
+	signalCtx = (void *)((char *)hal_cpuGetUserSP(ctx) - sizeof(*signalCtx));
+
+	/* changing sigmask and sleep shall be atomic - do it under lock (sigpost is done also under threads_common.spinlock) */
+	hal_spinlockSet(&threads_common.spinlock, &sc);
+	oldmask = thread->sigmask;
+	thread->sigmask = mask;
+
+	/* check for pending signals before sleep - with the new mask */
+	if (_threads_checkSignal(thread, thread->process, signalCtx, oldmask, SIG_SRC_SCALL) == 0) {
+		f = thread->process->sighandler;
+		hal_spinlockClear(&threads_common.spinlock, &sc);
+		hal_jmp(f, kstackTop, hal_cpuGetUserSP(signalCtx), 0, NULL);
+		/* no return */
+	}
+
+	/* FIXME: changing sigmask and sleep shall be atomic - introduce atomic unlock & sleep */
+	hal_spinlockClear(&threads_common.spinlock, &sc);
+	do {
+		ret = proc_threadSleep(1ULL << 52);
+	} while (ret != -EINTR);
+	hal_spinlockSet(&threads_common.spinlock, &sc);
```

Comment:
Use `proc_threadWaitInterruptible` on temp queue, so you don't have to let go of the spinlock